ext/mbstring/mbstring.c

```diff
@@ -4810,8 +4810,80 @@ static bool mb_fast_check_utf8_default(zend_string *str)
 
 	return true;
 # else
-	/* No SSE2 support; we might add generic UTF-8 specific validation code here later */
-	return php_mb_check_encoding(ZSTR_VAL(str), ZSTR_LEN(str), &mbfl_encoding_utf8);
+	/* This UTF-8 validation function is derived from PCRE2 */
+	size_t length = ZSTR_LEN(str);
+	/* Table of the number of extra bytes, indexed by the first byte masked with
+	0x3f. The highest number for a valid UTF-8 first byte is in fact 0x3d. */
+	static const uint8_t utf8_table[] = {
+		1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+		1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+		2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
+		3,3,3,3,3,3,3,3
+	};
+
+	for (; length > 0; p++) {
+		uint32_t d;
+		unsigned char c = *p;
+		length--;
+
+		if (c < 128) {
+			/* ASCII character */
+			continue;
+		}
+
+		if (c < 0xc0) {
+			/* Isolated 10xx xxxx byte */
+			return false;
+		}
+
+		if (c >= 0xf5) {
+			return false;
+		}
+
+		uint32_t ab = utf8_table[c & 0x3f]; /* Number of additional bytes (1-5) */
```

Comment:
The comment should be adjusted, now that we only handle up to 4 bytes.

Reply:
Very true! Thanks.